### DISCUSSION TITLE
Code-first gRPC

### DISCUSSION THREAD

**Discussion by flibbertigibbet-uk** [microsoft/aspire#2751] [General]
I am considering migrating a C# Core solution to Aspire and this solution currently uses a library that implements code-first gRPC [https://protobuf-net.github.io/protobuf-net.Grpc/gettingstarted](url) and [https://github.com/protobuf-net/protobuf-net.Grpc](url)

Has anyone been down this route?

I notice @JamesNK (co-author of protobuf-net.Grpc) has been active in this repository and previously discussed support of named gRPC end-points. [https://github.com/dotnet/aspire/issues/186](url) Is this delayed feature relevant to migrating protobuf-net.Grpc based application code to Aspire?

And before hitting the submit button I did some extra searching and found this open issue [https://github.com/dotnet/aspire/issues/898](url) which references the eShop sample. The sample code includes application level extension methods to support gRPC service discovery. I think I can just lift the gRPC extension method from the eShop code and replace:

`services.AddGrpcClient<TClient>`

with

`services.AddCodeFirstGrpcClient<TClient>`

**Comment by davidfowl** [comment]
This feature is a pure convenience over what is there today but AFAIK code first grpc should just work with service discovery. Is that what you are asking?

**Comment by azzageee** [comment]
You can do this by adapting the creator of the AddGrpcClient in the `AddGrpcServiceReference<TClient>` from the ServiceDefaults project.

This would look something like this:

```C#
public static IHttpClientBuilder AddCodeFirstGrpcReference<TClient>(this IServiceCollection services, string address, string? healthCheckName = null, HealthStatus failureStatus = default)
        where TClient : class
    {
        ArgumentNullException.ThrowIfNull(services);

        if (!Uri.IsWellFormedUriString(address, UriKind.Absolute))
        {
            throw new ArgumentException("Address must be a valid absolute URI.", nameof(address));
        }

        var uri = new Uri(address);
        var builder = services.AddGrpcClient<TClient>(o =>
        {
            o.Address = uri;
            o.Creator = invoker => invoker.CreateGrpcService<TClient>();
        });

        AddGrpcHealthChecks(services, uri, healthCheckName ?? $"{typeof(TClient).Name}-health", failureStatus);

        return builder;
    }
```

The key change here is with the creator adding `o.Creator = invoker => invoker.CreateGrpcService<TClient>();`
